# Uso de Pandas
En el desarrollo de este caso introduciremos las librerías pandas, numpy y matplotlib para el preprocesamiento, visualización de un conjunto de datos y el desarrollo de algunas medidas estadísticas descriptivas.

```{admonition} Pandas
Pandas es la librería aliada para el análisis de datos, con ella puede cargar fácilmente datos en formato CSV, Excel, SQL, entre otros. Además, permite realizar operaciones de filtrado, limpieza, transformación y análisis de datos. 
```


##  Introducción 

La revisión del comportamiento de los clientes en cualquier empresa es de vital importancia para el desarrollo de estrategias  tanto de mejoramiento de productos como de publicidad y toma de decisiones. Evaluaremos la información obtenida de un conjunto de registro obtenidos de una empresa de telecomunicaciones que cuenta con la revisión de las siguientes variables:

1. totrev: Ingresos del cliente
2. totmou: Total de minutos usados por el cliente
3. area: Área geográfica
4. creditcd: Indicador de tarjeta de crédito
5. eqpdays: Número de días (antigüedad) del equipo actual

Es importante mencionar que la empresa contiene una gran cantidad de clientes y hacer el análisis uno a uno no es lo  óptimo. Con ayuda de la programación y el análisis de datos, se logrará llegar a conclusiones rápidas y acertadas para encontrar relaciones que a simple vista no son tenidas en cuenta.

### ¿Qué se quiere identificar? 

Teniendo en cuenta la información de la base de datos se quiere identificar si hay una relación entre el total de minutos usados por el cliente con la compañía y el ingreso total de los clientes, es decir, Será posible afirmar que cuando una persona
tiene un mayor ingreso mayor entonces es mayor la utilización de minutos o todo lo contrario?


####  Insumos para tratar este problema


Teniendo en cuenta que la compañía tenia estándares para la entrega de la información, se dispone de un  conjunto de tablas en formato CSV ordenado de la siguiente manera:
La información de este conjunto de datos puede encontrase en [aquí](https://www.kaggle.com/abhinav89/telecom-customer?select=Telecom_customer+churn.csv)

### Objetivos

En este caso usted tendrá que cargar varias bases de datos, hará una exploración básica sobre la información y fusionará las distintas bases para tener una visión general del problema. La idea es que desarrolle las siguientes habilidades:

1. Manejo de la librería pandas de Python para cargar y leer datos; 
2. Ideas elementales para una útil transformación de los datos;
3. Construcción y presentación de argumentos válidos que le brindarán una solución para la pregunta expresada más arriba y llegar a una conclusión. 
4. Visualización de gráficas elementales que le permitirá entender mejor la información contenida en los datos.

#### Importación de paquetes


Una de las mejores opciones para trabajar con datos tabulares en Python es usar el módulo pandas. La librería `pandas` provee estructuras de datos, genera gráficos de alta calidad con `matplotlib` y se integra de buena forma con otras librerías que usan arrays de `numpy`.

Debemos revisar si la librería se encuentra en nuestro sistema usando el comando 

```python
!pip show librería
```


In [1]:
!pip show pandas
#!pip show numpy
#!pip show matplotlib

Name: pandas
Version: 2.1.3
Summary: Powerful data structures for data analysis, time series, and statistics
Home-page: https://pandas.pydata.org
Author: 
Author-email: The Pandas Development Team <pandas-dev@python.org>
License: BSD 3-Clause License

Copyright (c) 2008-2011, AQR Capital Management, LLC, Lambda Foundry, Inc. and PyData Development Team
All rights reserved.

Copyright (c) 2011-2023, Open source contributors.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of the copyright holder nor the names of its
  contributors may be u

En el caso de que la librería no se encuentre instalada ejecutamos el comando

```python
!pip install librería
```

Para importar las librerías usamos el comando 
```python
import librería as alias
```

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

### Eliminación de warnings

import warnings
warnings.filterwarnings('ignore')

Luego de que tener los paquetes activados dentro de nuestro ambiente de python podremos usar las fucnciones que se encuentren en la librería.

### Introducción a pandas

Poner información sobre pandas e indicar la diferencia entre series y dataframe

![](https://www.cdn.geeksforgeeks.org/wp-content/uploads/creating_dataframe1.png)

A continuación, mostramos el código a ejecutar para crear una ` Serie`

In [3]:
Serie1 = pd.Series(index=['Primero', 'Segundo', 'Tercero', 'Cuarto', 'Quinto'],
 name="Variable1", data=["A","B","C","D","E"]
)
Serie1

Primero    A
Segundo    B
Tercero    C
Cuarto     D
Quinto     E
Name: Variable1, dtype: object

In [4]:
Data_Frame = pd.DataFrame( 
index=['Primero', 'Segundo', 'Tercero', 'Cuarto', 'Quinto'],
    columns=["Variable2","Variable3"], 
    data=[['Hola', 12], ['Mi', 22],['Nombre',32.0],['es',42],['frailejon',52]]
)
Data_Frame

Variable2  Variable3
Primero       Hola       12.0
Segundo         Mi       22.0
Tercero     Nombre       32.0
Cuarto          es       42.0
Quinto   frailejon       52.0

Para revisar el tipo de objeto que contiene nuestro DataFrame podemos ejecutar

In [5]:
Data_Frame.dtypes
#Serie1.dtypes

Variable2     object
Variable3    float64
dtype: object

Extraer datos sobre la base:

Si quiero conocer una columna de la base simplemente escribo `DataFrame['nombre_columna']`

In [6]:
Data_Frame['Variable3']

Primero    12.0
Segundo    22.0
Tercero    32.0
Cuarto     42.0
Quinto     52.0
Name: Variable3, dtype: float64

Para filtrar datos

In [7]:
DF_filtrado=Data_Frame[Data_Frame['Variable2']=='Hola'][['Variable3']]

In [8]:
DF_filtrado

Variable3
Primero       12.0

Para crear nuevas variables

In [9]:
Data_Frame['Variable1']=[1,2,3,4,5]

In [10]:
Data_Frame

Variable2  Variable3  Variable1
Primero       Hola       12.0          1
Segundo         Mi       22.0          2
Tercero     Nombre       32.0          3
Cuarto          es       42.0          4
Quinto   frailejon       52.0          5

Si quiero ver varias columnas de la base de datos, utilizo doble paréntesis cuadrado:

`DataFrame[['columna1','columna2']]`

In [11]:
Data_Frame[['Variable1','Variable2','Variable1']]

Variable1  Variable2  Variable1
Primero          1       Hola          1
Segundo          2         Mi          2
Tercero          3     Nombre          3
Cuarto           4         es          4
Quinto           5  frailejon          5

Si quiero ver datos por registros utilizo loc:

In [12]:
Data_Frame.iloc[[0,2,4]]

Variable2  Variable3  Variable1
Primero       Hola       12.0          1
Tercero     Nombre       32.0          3
Quinto   frailejon       52.0          5

In [13]:
Data_Frame[Data_Frame['Variable2']=='frailejon']

Variable2  Variable3  Variable1
Quinto  frailejon       52.0          5

In [14]:
Data_Frame['Variable2']['Quinto']='Frailejon'

```{}
En las siguientes celdas se muestran una serie de procedimientos, intenta identificar que se esta haciendo.

In [15]:

for i in Data_Frame.index:
    print(Data_Frame.loc[i])

Variable2    Hola
Variable3    12.0
Variable1       1
Name: Primero, dtype: object
Variable2      Mi
Variable3    22.0
Variable1       2
Name: Segundo, dtype: object
Variable2    Nombre
Variable3      32.0
Variable1         3
Name: Tercero, dtype: object
Variable2      es
Variable3    42.0
Variable1       4
Name: Cuarto, dtype: object
Variable2    Frailejon
Variable3         52.0
Variable1            5
Name: Quinto, dtype: object


In [16]:
Data_Frame['Variable4']=['A','A','B','B','A']
Data_Frame['Variable3']=[12.0,22,32,42,52]

In [17]:
Data_Frame['VariableAux']=2*Data_Frame[Data_Frame['Variable4']=='A']['Variable3']
Data_Frame['VariableAux2']=Data_Frame[Data_Frame['Variable4']=='B']['Variable3']/2
Data_Frame['VariableAux']=Data_Frame['VariableAux'].fillna(0)
Data_Frame['VariableAux2']=Data_Frame['VariableAux2'].fillna(0)

In [18]:
Data_Frame['Variable3']=Data_Frame['VariableAux']+Data_Frame['VariableAux2']

In [19]:
Data_Frame

Variable2  Variable3  Variable1 Variable4  VariableAux  VariableAux2
Primero       Hola       24.0          1         A         24.0           0.0
Segundo         Mi       44.0          2         A         44.0           0.0
Tercero     Nombre       16.0          3         B          0.0          16.0
Cuarto          es       21.0          4         B          0.0          21.0
Quinto   Frailejon      104.0          5         A        104.0           0.0

### Carga de información en `pandas`

Para nuestro  ejercicio usaremos los siguientes conjuntos de datos alojados en [este enlace](https://github.com/Izainea/seminario-de-programacion/Cuadernos/datos_01):

1. NEW ENGLAND AREA.csv
2. ATLANTIC SOUTH AREA.csv
3. NEW YORK CITY AREA.csv
4. CHICAGO AREA.csv
5. NORTH FLORIDA AREA.csv
6. DC-MARYLAND-VIRGINIA AREA.csv
7. GREAT LAKES AREA.csv
8. NORTHWEST-ROCKY MOUNTAIN AREA.csv
9. MIDWEST AREA.csv
10. SOUTHWEST AREA.csv
11. LOS ANGELES AREA.csv
12. HOUSTON AREA.csv
13. CALIFORNIA NORTH AREA.csv
14. CENTRAL-SOUTH TEXAS AREA.csv
15. DALLAS AREA.csv
16. PHILADELPHIA AREA.csv
17. TENNESSEE AREA.csv
18. OHIO AREA.csv
19. SOUTH FLORIDA AREA.csv

Cada uno de estos conjuntos de datos contiene la información de:

1. *mou_Mean:* Número medio de minutos de uso mensuales
2. *custcare_Mean:* Número medio de llamadas de atención al cliente
3. *area:*  Área geografica
4. *creditcd:* Indicador de tarjeta de crédito
5. *eqpdays:* Número de días (antigüedad) del equipo actual

Además, se incluye _Credito_dias.csv_ con la información de 2 variables adicionales para los registros las cuales son:

1. _creditcd:_ Indicador de tarjeta de crédito
2. _eqpdays:_ Número de días (antigüedad) del equipo actual

Revisaremos inicialmente los datos contenidos en _Credito_dias.csv_ para revisar algunas funciones interesantes de `pandas`, usaremos el comando 
```python
pd.read_csv("archivo.csv")
```
Podemos definir el número de registros a imprimir usando
```python
pd.options.display.min_rows=10
````

```{hint} Carga de datos
Si usa Google Colab o anaconda procure guardar todo lo que descarga en la carpeta de `datos_01` al mismo nivel de su cuaderno de practica, si utiliza este descargado desde [aquí](https://github.com/izainea/seminario-de.programacion/Cuadernos/Minutos y segundos.ipynb)

La estructura de la carpeta de trabajo debe ser así:

```
Minutos y segundos.ipynb
datos_01
- NEW ENGLAND AREA.csv
- ATLANTIC SOUTH AREA.csv
- NEW YORK CITY AREA.csv
...

```



Usando el paquete `os` identificaremos donde están nuestros archivos:

In [20]:
import os

#os.getcwd()

In [21]:
os.listdir()

['.ipynb_checkpoints',
 'Aprendizaje Automático.ipynb',
 'bank_marketing.csv',
 'Base Grande.csv',
 'Clase 2',
 'datos_01',
 'Ejercicio K Means.ipynb',
 'Estadisticos de Validación.ipynb',
 'Graficos.ipynb',
 'ifood_df.csv',
 'images',
 'Intro a Python y Pandas.md',
 'introduccion.md',
 'Manipulación, Limpieza y Exploración de Datos.ipynb',
 'Medidas Estadisticas.ipynb',
 'Minutos y segundos.ipynb',
 'Modelos de Clasificación I.ipynb',
 'Modelos de Clasificación II.ipynb',
 'Visualización en Python.ipynb',
 'Árboles de decisión.ipynb']

Como asumimos que todo se guardó en la carpeta `datos_01`, tenemos:

In [22]:
os.listdir('datos_01')

['ATLANTIC SOUTH AREA.csv',
 'CALIFORNIA NORTH AREA.csv',
 'CENTRAL-SOUTH TEXAS AREA.csv',
 'CHICAGO AREA.csv',
 'Credito_dias.csv',
 'DALLAS AREA.csv',
 'DC-MARYLAND-VIRGINIA AREA.csv',
 'GREAT LAKES AREA.csv',
 'HOUSTON AREA.csv',
 'LOS ANGELES AREA.csv',
 'MIDWEST AREA.csv',
 'NEW ENGLAND AREA.csv',
 'NEW YORK CITY AREA.csv',
 'NORTH FLORIDA AREA.csv',
 'NORTHWEST-ROCKY MOUNTAIN AREA.csv',
 'OHIO AREA.csv',
 'PHILADELPHIA AREA.csv',
 'SOUTH FLORIDA AREA.csv',
 'SOUTHWEST AREA.csv',
 'TENNESSEE AREA.csv']

Como sabemos donde esta el archivo, lo cargamos:

In [23]:
import pandas as pd
df=pd.read_csv("datos_01/Credito_dias.csv", sep=",",index_col="Customer_ID")
df

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1069232      13552.0  1172.25        N       83
1031705       4152.0   577.39        Y      675
1089806       1863.0   294.26        Y      218
1076695       4155.0   433.16        N      222
...              ...      ...      ...      ...
1027614       2570.0   484.31        Y      611
1098422       2105.0   583.36        Y     1243
1096497       3690.0   294.29        Y      212
1087248       3333.0   233.08        Y      290
1099003       5025.0  1436.49        Y     1029

[10000 rows x 4 columns]

El conjunto de datos se encuentra en el `DataFrame` `df`

sobre el objeto creado podemos usar  algunas funciones  para darnos una idea del comportamiento de la información:
```python
df.head(n)  # imprime los primeros n registros del DataFrame
df.tail(n)  # imprime los últimos n registros del DataFrame 
df.shape    # imprime el numero de columnas y filas del DataFrame
df.columns  # imprime el nombre de las columnas del DataFrame
df.index    # imprime el indice de los registros del DataFrame
df.dtypes   # imprime el tipo de cada una de las columnnas del DataFrame
df.sample(n)# imprime una muestra aleatoria de n registros en el DataFrame
```

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10000 entries, 1082423 to 1099003
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   totmou    10000 non-null  float64
 1   totrev    10000 non-null  float64
 2   creditcd  9848 non-null   object 
 3   eqpdays   10000 non-null  int64  
dtypes: float64(2), int64(1), object(1)
memory usage: 390.6+ KB


Veamos información relevante de la variable `totmou`

In [25]:
df['totmou']

Customer_ID
1082423     7269.0
1069232    13552.0
1031705     4152.0
1089806     1863.0
1076695     4155.0
            ...   
1027614     2570.0
1098422     2105.0
1096497     3690.0
1087248     3333.0
1099003     5025.0
Name: totmou, Length: 10000, dtype: float64

In [26]:
df['totmou'].mean()

7692.8048656727005

In [27]:
df.describe(include='all')

totmou        totrev creditcd       eqpdays
count    10000.000000  10000.000000     9848  10000.000000
unique            NaN           NaN        2           NaN
top               NaN           NaN        Y           NaN
freq              NaN           NaN     6796           NaN
mean      7692.804866   1042.133807      NaN    386.794500
std       8657.917883    857.144895      NaN    252.720196
min          0.000000     11.100000      NaN     -5.000000
25%       2576.000000    526.787500      NaN    210.000000
50%       5262.000000    820.235000      NaN    338.000000
75%       9823.250000   1272.975000      NaN    522.000000
max     163264.176700  13426.600000      NaN   1823.000000

In [28]:
display("primeros 3 registros",df.head(3))
display("3 últimos registros",df.tail(3))
display("Columnas:",df.columns)
display("Índices:",df.index)
display("Tipos de registros en el DataFrame",df.dtypes)
display("Muestra aleatoria de 3 registros",df.sample(3))
display("dimensión",df.shape)

'primeros 3 registros'

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1069232      13552.0  1172.25        N       83
1031705       4152.0   577.39        Y      675

'3 últimos registros'

totmou   totrev creditcd  eqpdays
Customer_ID                                   
1096497      3690.0   294.29        Y      212
1087248      3333.0   233.08        Y      290
1099003      5025.0  1436.49        Y     1029

'Columnas:'

Index(['totmou', 'totrev', 'creditcd', 'eqpdays'], dtype='object')

'Índices:'

Index([1082423, 1069232, 1031705, 1089806, 1076695, 1091652, 1043846, 1088967,
       1045239, 1046584,
       ...
       1087145, 1002540, 1044082, 1083458, 1093976, 1027614, 1098422, 1096497,
       1087248, 1099003],
      dtype='int64', name='Customer_ID', length=10000)

'Tipos de registros en el DataFrame'

totmou      float64
totrev      float64
creditcd     object
eqpdays       int64
dtype: object

'Muestra aleatoria de 3 registros'

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1027346      14114.0  1642.39        N      651
1055309         10.0   212.64        Y      456
1035677      18525.0  2685.81        Y      641

'dimensión'

(10000, 4)

In [29]:
df.creditcd

Customer_ID
1082423    Y
1069232    N
1031705    Y
1089806    Y
1076695    N
          ..
1027614    Y
1098422    Y
1096497    Y
1087248    Y
1099003    Y
Name: creditcd, Length: 10000, dtype: object

Este conjunto de datos corresponde a 10000 registros y 2 variables

### Manejo de datos duplicados 

os dataframes tienen la posibilidad de detectar las filas duplicadas con la función: 
```
df.duplicated({columns})
```
Si no especifica {columns}, se busacarán duplicados teniendo en cuenta todas las columnas. Una vez detectados el dtaframe tiene un función para eliminar filas duplicadas, haciendo:

```
df = df.drop_duplicates()
```
También es posible eliminar filas que duplican solo algunos campos. Con el fin de dejar solo una ocurrencia.

```
df.drop_duplicates(['nombre', 'apellido', 'cédula'], keep='last')
```
En este caso, como no todas las colomnas son iguales se conserva solo la última ocurrencia.

### Manejo de datos perdidos o esperados
Son aquellos datos flatantes en la tabla que por alguna razón o error no se encuentran, o parecen con algún valor o etiqueta de no válido. 
Para manipular datos perdidos use la explicación dada el siguiente enlace: [DATOS PERDIDOS](http://ligdigonzalez.com/manipulando-datos-perdidos-en-python/)
recuerde que para reemplazar datos perdidos en un dataframe $df$ se usa la función:
```python
promedio = df['totrev'].mean()
df['nota'].replace(np.nan, promedio)
```
en donde la opción 'promedio' busca el valor promedio de la columna (datos numéricos) o la moda (datos categóricos) para reemplazar el valor no válido.


In [30]:
df.isnull().any(axis=0)

totmou      False
totrev      False
creditcd     True
eqpdays     False
dtype: bool

In [31]:
perd=df.isnull().any(axis=1)
perd

Customer_ID
1082423    False
1069232    False
1031705    False
1089806    False
1076695    False
           ...  
1027614    False
1098422    False
1096497    False
1087248    False
1099003    False
Length: 10000, dtype: bool

In [32]:
range(len(df))

range(0, 10000)

In [33]:
df.index

Index([1082423, 1069232, 1031705, 1089806, 1076695, 1091652, 1043846, 1088967,
       1045239, 1046584,
       ...
       1087145, 1002540, 1044082, 1083458, 1093976, 1027614, 1098422, 1096497,
       1087248, 1099003],
      dtype='int64', name='Customer_ID', length=10000)

In [34]:
for i in df.index[:10]:
    print(i)

1082423
1069232
1031705
1089806
1076695
1091652
1043846
1088967
1045239
1046584


In [35]:
df[perd]

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1072218       5684.0   857.65      NaN      244
1066597        925.0   373.12      NaN      344
1023352      14466.0  2243.76      NaN      387
1045521       2504.0   756.79      NaN      495
1001813      76260.0  5526.11      NaN      361
...              ...      ...      ...      ...
1050795      15591.0  1330.71      NaN      447
1096985       3011.0   522.14      NaN       74
1025203        853.0   650.34      NaN      780
1099258      33660.0  4247.85      NaN      206
1026149       1955.0   606.08      NaN      770

[152 rows x 4 columns]

In [36]:
perdidos = df[df.isnull().any(axis=1)]
perdidos

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1072218       5684.0   857.65      NaN      244
1066597        925.0   373.12      NaN      344
1023352      14466.0  2243.76      NaN      387
1045521       2504.0   756.79      NaN      495
1001813      76260.0  5526.11      NaN      361
...              ...      ...      ...      ...
1050795      15591.0  1330.71      NaN      447
1096985       3011.0   522.14      NaN       74
1025203        853.0   650.34      NaN      780
1099258      33660.0  4247.85      NaN      206
1026149       1955.0   606.08      NaN      770

[152 rows x 4 columns]

In [37]:
df['creditcd'].value_counts(dropna=False)

creditcd
Y      6796
N      3052
NaN     152
Name: count, dtype: int64

In [38]:
promedio = df['totrev'].mean()
df['totrev']=df['totrev'].replace(np.nan, promedio)

También es recomendado usar la funcion `fillna()`

In [39]:
df['creditcd'].fillna("Y",inplace=True)

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10000 entries, 1082423 to 1099003
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   totmou    10000 non-null  float64
 1   totrev    10000 non-null  float64
 2   creditcd  10000 non-null  object 
 3   eqpdays   10000 non-null  int64  
dtypes: float64(2), int64(1), object(1)
memory usage: 390.6+ KB


In [41]:
df.isnull().any(axis=0)

totmou      False
totrev      False
creditcd    False
eqpdays     False
dtype: bool

In [42]:
df

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1069232      13552.0  1172.25        N       83
1031705       4152.0   577.39        Y      675
1089806       1863.0   294.26        Y      218
1076695       4155.0   433.16        N      222
...              ...      ...      ...      ...
1027614       2570.0   484.31        Y      611
1098422       2105.0   583.36        Y     1243
1096497       3690.0   294.29        Y      212
1087248       3333.0   233.08        Y      290
1099003       5025.0  1436.49        Y     1029

[10000 rows x 4 columns]

In [43]:
df['creditcd'].value_counts(dropna=False)

creditcd
Y    6948
N    3052
Name: count, dtype: int64



### Cálculo de estadísticas resumen

Las medidas estadísticas resumen son de gran ayuda para comprender como es el comportamiento de la distribución de los datos, los objetos `DataFrame` ofrecen facilidades para hacer el cálculo de algunas estadísticas sobre cada una de las variables.

```python
df["var"].value_counts() # Encuentra la frecuencia de las categorias de una columna (la columna debe ser cualitativa)
df["var"].min()          # Encuentra el mínimo de una variable cuantitativa
df["var"].median()       # Encuentra el mínimo de una variable cuantitativa
df["var"].mean()         # Encuentra el promedio de una variable cuantitativa
df["var"].max()          # Encuentra el máximo de una variable cuantitativa
df["var"].quantile()     # Encuentra los cuantiles de una variable cuantitativa
```


In [44]:
df["creditcd"].value_counts()

creditcd
Y    6948
N    3052
Name: count, dtype: int64

In [45]:
df["creditcd"].count()

10000

In [46]:
df["eqpdays"].min()

-5

In [47]:
df["eqpdays"]=[0 if i<0 else i for i in df["eqpdays"] ]

In [48]:
df["eqpdays"].min()

0

La exploración del anterior resultado puede mostrar inconsistencias  en los registros y nos ayudará a tomar decisiones sobre algunos registros, el valor mínimo de la variable `eqpdays` (Número de días (antigüedad) del equipo actual) debería ser estrictamente positivo y estos pueden ser errores de ingreso.

In [49]:
df["eqpdays"].max()

1823

In [50]:
df["eqpdays"].median()

338.0

In [51]:
df["eqpdays"].mean()

386.7992

In [52]:
df["eqpdays"].quantile(0.25)

210.0

In [53]:
df["eqpdays"].quantile(0.75)

522.0

Estas estadísticas pueden ser resumidas con el comando:
```python
df["var"].describe()  # Estadísticas resumen principales
```

In [54]:
df["eqpdays"].describe()

count    10000.00000
mean       386.79920
std        252.71297
min          0.00000
25%        210.00000
50%        338.00000
75%        522.00000
max       1823.00000
Name: eqpdays, dtype: float64

In [55]:
df['creditcd'].describe()

count     10000
unique        2
top           Y
freq       6948
Name: creditcd, dtype: object

In [56]:
df.describe()

totmou        totrev      eqpdays
count   10000.000000  10000.000000  10000.00000
mean     7692.804866   1042.133807    386.79920
std      8657.917883    857.144895    252.71297
min         0.000000     11.100000      0.00000
25%      2576.000000    526.787500    210.00000
50%      5262.000000    820.235000    338.00000
75%      9823.250000   1272.975000    522.00000
max    163264.176700  13426.600000   1823.00000

In [57]:
df.describe(include='all')

totmou        totrev creditcd      eqpdays
count    10000.000000  10000.000000    10000  10000.00000
unique            NaN           NaN        2          NaN
top               NaN           NaN        Y          NaN
freq              NaN           NaN     6948          NaN
mean      7692.804866   1042.133807      NaN    386.79920
std       8657.917883    857.144895      NaN    252.71297
min          0.000000     11.100000      NaN      0.00000
25%       2576.000000    526.787500      NaN    210.00000
50%       5262.000000    820.235000      NaN    338.00000
75%       9823.250000   1272.975000      NaN    522.00000
max     163264.176700  13426.600000      NaN   1823.00000

### Agregar información de múltiples tablas

Hemos desarrollado un breve resumen estadístico solamente usando la base `Credito_dias.csv`. Realizaremos la combinación de las 19 áreas que nos reportan en los archivos csv. Una forma de lograr esta tarea de agregación es usar el método pd.concat() de pandas. Una entrada en este método puede ser una lista de DataFrames que quiera concatenar. Usaremos un ciclo  `for` sobre cada uno de los archivos de áreas para 


In [58]:
df.columns
df.dtypes

totmou      float64
totrev      float64
creditcd     object
eqpdays       int64
dtype: object

### Ejercicio 1
Determinar los percentiles 25, 50 y 75 para las columnas ttomou, totrev, eqpdays

In [59]:
#Espacio para el desarrollo del ejercicio
df["totmou"].describe()

count     10000.000000
mean       7692.804866
std        8657.917883
min           0.000000
25%        2576.000000
50%        5262.000000
75%        9823.250000
max      163264.176700
Name: totmou, dtype: float64

In [60]:
df["totrev"].describe()

count    10000.000000
mean      1042.133807
std        857.144895
min         11.100000
25%        526.787500
50%        820.235000
75%       1272.975000
max      13426.600000
Name: totrev, dtype: float64

In [61]:
df.describe(include='all')

totmou        totrev creditcd      eqpdays
count    10000.000000  10000.000000    10000  10000.00000
unique            NaN           NaN        2          NaN
top               NaN           NaN        Y          NaN
freq              NaN           NaN     6948          NaN
mean      7692.804866   1042.133807      NaN    386.79920
std       8657.917883    857.144895      NaN    252.71297
min          0.000000     11.100000      NaN      0.00000
25%       2576.000000    526.787500      NaN    210.00000
50%       5262.000000    820.235000      NaN    338.00000
75%       9823.250000   1272.975000      NaN    522.00000
max     163264.176700  13426.600000      NaN   1823.00000

In [62]:
areas=list(os.listdir("datos_01"))
try:
    areas.remove('.ipynb_checkpoints')
except:
    pass

try:
    areas.remove('Credito_dias.csv')
except:
    pass
areas=[i for i in areas if i.count('.csv')>=1 ]
areas

['ATLANTIC SOUTH AREA.csv',
 'CALIFORNIA NORTH AREA.csv',
 'CENTRAL-SOUTH TEXAS AREA.csv',
 'CHICAGO AREA.csv',
 'DALLAS AREA.csv',
 'DC-MARYLAND-VIRGINIA AREA.csv',
 'GREAT LAKES AREA.csv',
 'HOUSTON AREA.csv',
 'LOS ANGELES AREA.csv',
 'MIDWEST AREA.csv',
 'NEW ENGLAND AREA.csv',
 'NEW YORK CITY AREA.csv',
 'NORTH FLORIDA AREA.csv',
 'NORTHWEST-ROCKY MOUNTAIN AREA.csv',
 'OHIO AREA.csv',
 'PHILADELPHIA AREA.csv',
 'SOUTH FLORIDA AREA.csv',
 'SOUTHWEST AREA.csv',
 'TENNESSEE AREA.csv']

In [63]:
print("Definición de los símbolos de las acciones")

lista_de_df = []
# Bucle sobre los símbolos
#print(" --- Inicie el bucle sobre los símbolos --- ")
for i in areas:
    ## ruta completa
    i='datos_01/'+i
    #print("Procesando el símbolo: " + i)
    temp_df = pd.read_csv(i,index_col="Customer_ID")
    temp_df["area"] = i
    lista_de_df.append(temp_df)
    # Usando un salto de línea al final de esta cadena de caracteres por estética
   # print(" --- Bucle completo sobre los símbolos --- \n")
    # Combinando en un solo DataFrame usando el concat
print("Agregando los datos")
agr_df = pd.concat(lista_de_df, axis=0)
print(agr_df.shape)
print("Cabeza del DataFrame agr_df: ")
agr_df.head()

Definición de los símbolos de las acciones
Agregando los datos
(9997, 4)
Cabeza del DataFrame agr_df: 


custcare_Mean new_cell                              area creditcd
Customer_ID                                                                   
1069232           3.666667        U  datos_01/ATLANTIC SOUTH AREA.csv        N
1043846           0.000000        U  datos_01/ATLANTIC SOUTH AREA.csv        Y
1008539           0.000000        N  datos_01/ATLANTIC SOUTH AREA.csv        Y
1068315           0.000000        U  datos_01/ATLANTIC SOUTH AREA.csv        N
1074220           1.666667        U  datos_01/ATLANTIC SOUTH AREA.csv        N

In [64]:
texto=r"datos_01/"
agr_df['area']=agr_df['area'].str.replace(texto , '').str.replace(" AREA.csv","")

In [65]:
agr_df

custcare_Mean new_cell            area creditcd
Customer_ID                                                 
1069232           3.666667        U  ATLANTIC SOUTH        N
1043846           0.000000        U  ATLANTIC SOUTH        Y
1008539           0.000000        N  ATLANTIC SOUTH        Y
1068315           0.000000        U  ATLANTIC SOUTH        N
1074220           1.666667        U  ATLANTIC SOUTH        N
...                    ...      ...             ...      ...
1093180           3.000000        U       TENNESSEE        Y
1018795           0.000000        U       TENNESSEE        Y
1055110           9.666667        U       TENNESSEE        N
1078247           0.333333        U       TENNESSEE        Y
1061746           0.000000        U       TENNESSEE        Y

[9997 rows x 4 columns]

In [66]:
agr_df.sort_index(axis=0,inplace=True)
agr_df.to_csv('Base Grande.csv')

In [67]:
df

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1069232      13552.0  1172.25        N       83
1031705       4152.0   577.39        Y      675
1089806       1863.0   294.26        Y      218
1076695       4155.0   433.16        N      222
...              ...      ...      ...      ...
1027614       2570.0   484.31        Y      611
1098422       2105.0   583.36        Y     1243
1096497       3690.0   294.29        Y      212
1087248       3333.0   233.08        Y      290
1099003       5025.0  1436.49        Y     1029

[10000 rows x 4 columns]

In [68]:
agr_df

custcare_Mean new_cell                      area creditcd
Customer_ID                                                           
1000014           0.000000        Y       CENTRAL-SOUTH TEXAS        N
1000016           1.000000        N                    DALLAS        Y
1000021           0.333333        U                 TENNESSEE        Y
1000023           0.000000        N  NORTHWEST-ROCKY MOUNTAIN        N
1000029           0.000000        N               NEW ENGLAND        Y
...                    ...      ...                       ...      ...
1099907          17.333333        U               LOS ANGELES        Y
1099932           0.000000        U               LOS ANGELES        Y
1099963           0.000000        U                 SOUTHWEST        Y
1099981           0.333333        U                 SOUTHWEST      NaN
1099992           0.000000        Y               LOS ANGELES        Y

[9997 rows x 4 columns]

La ejecución nos da como resultado una base de datos de 9997 registros correspondiente a la concatenación de todas las áreas. Se debe notar que nuestro conjunto de datos también es indexado por la columna `Customer_ID` y que la dimensión de los registros de está nueva base es menor a la de  `Creditos_dias.csv`.

 ## Cruce de información

El cruce de información es una operación esencial en el análisis de los datos. Usualmente, la posibilidad de realizar el cruce de dos o más fuentes de información es el primer paso para responder preguntas complejas acerca de los datos. Existen muchas formas de cruzar o combinar tablas de datos. El cruce de estos datos siempre tiene un propósito. Por tanto, se debe realizar un análisis de lo que se requiere antes de iniciar. Para esto, lo primero es conocer bien las fuentes de información que se van a relacionar. 

### Uniones (Joins)
Usando la librería pandas se pueden realizar 4 tipos de uniones para cruzar información entre dos fuentes de datos, ver Figura abajo. 
* INNER JOIN: este tipo de cruce incluye solo los resgistros que coincieden en las dos tablas
* LEFT JOIN:  en este tipo de cruce se incluyen todos los registros de la primera tabla aunque no aparezcan coincidencias en la segunda. Cuando no hay coincidencias en la segunda tabla los campos correspondientes a esta aparecerán como nulos.
* RIGHT JOIN: en este tipo de cruce se incluyen todos los registros de la segunda tabla aunque no aparezcan coincidencias en la primera. Cuando no hay coincidencias en la primer tabla los campos correspondientes a esta aparecerán como nulos.
* OUTER JOIN O FULL JOIN: combina todos los registros de ambas tablas aunque no existan coincidencias en la otra tabla. 

![Los tipos de uniones que se pueden ejecutar en pandas](https://letsdobigdata.files.wordpress.com/2016/03/joins.png)

La combinación a usar depende del análisis que se requiera con los datos. Lo más común es realizar un cruce de tipo Inner Join, debido a que muestra los registros cuando aparecen relacionados en ambas tablas. Sin embargo, en 
ocasiones se quiere saber que pasa con todos los registros de la primera tabla aunque no se tengan coincidencias. Alternativamente, quizá, lo menos usual es hacer combinaciones de tipo RIGHT o OUTER. Generalmente, estas son usadas cuando se quiere saber por qué no se tienen coincidencias, o cuáles registros no cumplen con la relación especificada.

Ya que desconocemos cuales son los registros perdidos en los 2 dataframes que hemos construido y es necesario para continuar nuestro preprocesamiento el unir la información.

In [69]:
agr_df=agr_df.reset_index()
agr_df

Customer_ID  custcare_Mean new_cell                      area creditcd
0         1000014       0.000000        Y       CENTRAL-SOUTH TEXAS        N
1         1000016       1.000000        N                    DALLAS        Y
2         1000021       0.333333        U                 TENNESSEE        Y
3         1000023       0.000000        N  NORTHWEST-ROCKY MOUNTAIN        N
4         1000029       0.000000        N               NEW ENGLAND        Y
...           ...            ...      ...                       ...      ...
9992      1099907      17.333333        U               LOS ANGELES        Y
9993      1099932       0.000000        U               LOS ANGELES        Y
9994      1099963       0.000000        U                 SOUTHWEST        Y
9995      1099981       0.333333        U                 SOUTHWEST      NaN
9996      1099992       0.000000        Y               LOS ANGELES        Y

[9997 rows x 5 columns]

In [70]:
df=df.reset_index()


In [71]:
df2=df.copy()

In [72]:
df

Customer_ID   totmou   totrev creditcd  eqpdays
0         1082423   7269.0   906.58        Y      371
1         1069232  13552.0  1172.25        N       83
2         1031705   4152.0   577.39        Y      675
3         1089806   1863.0   294.26        Y      218
4         1076695   4155.0   433.16        N      222
...           ...      ...      ...      ...      ...
9995      1027614   2570.0   484.31        Y      611
9996      1098422   2105.0   583.36        Y     1243
9997      1096497   3690.0   294.29        Y      212
9998      1087248   3333.0   233.08        Y      290
9999      1099003   5025.0  1436.49        Y     1029

[10000 rows x 5 columns]

In [73]:
#for i in df2.index[:10]:
#    df['Customer_ID'][i]=df['Customer_ID'][i]+10

In [74]:
df.head(10)

Customer_ID   totmou   totrev creditcd  eqpdays
0      1082423   7269.0   906.58        Y      371
1      1069232  13552.0  1172.25        N       83
2      1031705   4152.0   577.39        Y      675
3      1089806   1863.0   294.26        Y      218
4      1076695   4155.0   433.16        N      222
5      1091652   5441.0   578.80        Y       68
6      1043846   7294.0  1495.18        Y      460
7      1088967   4395.0   668.23        N      333
8      1045239  11544.0  1012.64        Y      498
9      1046584   3751.0  1199.89        Y      592

In [75]:
DF_Derecha=pd.merge(df,agr_df,how='right',left_on='Customer_ID',right_on='Customer_ID')
DF_Derecha

Customer_ID       totmou   totrev creditcd_x  eqpdays  custcare_Mean  \
0         1000014    642.00000  1370.94          N      354       0.000000   
1         1000016  15226.03000  2311.08          Y      697       1.000000   
2         1000021  79406.01667  6751.70          Y      143       0.333333   
3         1000023  64174.80667  7419.73          N      350       0.000000   
4         1000029  42715.00000  3829.33          Y      183       0.000000   
...           ...          ...      ...        ...      ...            ...   
9992      1099907  17843.00000  1457.01          Y      184      17.333333   
9993      1099932  14079.00000  1538.55          Y      275       0.000000   
9994      1099963   9722.00000  1559.16          Y      946       0.000000   
9995      1099981   7248.00000  1417.30          Y      882       0.333333   
9996      1099992   7338.00000  1198.48          Y      773       0.000000   

     new_cell                      area creditcd_y  
0           Y       CENTRAL-SOUTH TEXAS          N  
1           N                    DALLAS          Y  
2           U                 TENNESSEE          Y  
3           N  NORTHWEST-ROCKY MOUNTAIN          N  
4           N               NEW ENGLAND          Y  
...       ...                       ...        ...  
9992        U               LOS ANGELES          Y  
9993        U               LOS ANGELES          Y  
9994        U                 SOUTHWEST          Y  
9995        U                 SOUTHWEST        NaN  
9996        Y               LOS ANGELES          Y  

[9997 rows x 9 columns]

In [76]:
DF_Derecha[DF_Derecha['totmou'].isna()]

Empty DataFrame
Columns: [Customer_ID, totmou, totrev, creditcd_x, eqpdays, custcare_Mean, new_cell, area, creditcd_y]
Index: []

In [77]:
df.head(10)

Customer_ID   totmou   totrev creditcd  eqpdays
0      1082423   7269.0   906.58        Y      371
1      1069232  13552.0  1172.25        N       83
2      1031705   4152.0   577.39        Y      675
3      1089806   1863.0   294.26        Y      218
4      1076695   4155.0   433.16        N      222
5      1091652   5441.0   578.80        Y       68
6      1043846   7294.0  1495.18        Y      460
7      1088967   4395.0   668.23        N      333
8      1045239  11544.0  1012.64        Y      498
9      1046584   3751.0  1199.89        Y      592

In [78]:
DF_Izq=pd.merge(df,agr_df,how='left',left_on='Customer_ID',right_on='Customer_ID')
DF_Izq

Customer_ID   totmou   totrev creditcd_x  eqpdays  custcare_Mean  \
0         1082423   7269.0   906.58          Y      371      14.333333   
1         1069232  13552.0  1172.25          N       83       3.666667   
2         1031705   4152.0   577.39          Y      675       3.000000   
3         1089806   1863.0   294.26          Y      218      11.000000   
4         1076695   4155.0   433.16          N      222       0.000000   
...           ...      ...      ...        ...      ...            ...   
9995      1027614   2570.0   484.31          Y      611       0.666667   
9996      1098422   2105.0   583.36          Y     1243       0.000000   
9997      1096497   3690.0   294.29          Y      212       0.666667   
9998      1087248   3333.0   233.08          Y      290       0.000000   
9999      1099003   5025.0  1436.49          Y     1029       0.333333   

     new_cell              area creditcd_y  
0           U       NEW ENGLAND          Y  
1           U    ATLANTIC SOUTH          N  
2           U     NEW YORK CITY          Y  
3           U           CHICAGO          Y  
4           U     NORTH FLORIDA          N  
...       ...               ...        ...  
9995        U           CHICAGO          Y  
9996        U         SOUTHWEST          Y  
9997        Y       NEW ENGLAND          Y  
9998        U  CALIFORNIA NORTH          Y  
9999        U       LOS ANGELES          Y  

[10000 rows x 9 columns]

In [79]:
DF_Izq['area'].isna().sum()

3

In [80]:
DF_Full=pd.merge(df,agr_df,how='outer',left_on='Customer_ID',right_on='Customer_ID')
DF_Full

Customer_ID   totmou   totrev creditcd_x  eqpdays  custcare_Mean  \
0         1082423   7269.0   906.58          Y      371      14.333333   
1         1069232  13552.0  1172.25          N       83       3.666667   
2         1031705   4152.0   577.39          Y      675       3.000000   
3         1089806   1863.0   294.26          Y      218      11.000000   
4         1076695   4155.0   433.16          N      222       0.000000   
...           ...      ...      ...        ...      ...            ...   
9995      1027614   2570.0   484.31          Y      611       0.666667   
9996      1098422   2105.0   583.36          Y     1243       0.000000   
9997      1096497   3690.0   294.29          Y      212       0.666667   
9998      1087248   3333.0   233.08          Y      290       0.000000   
9999      1099003   5025.0  1436.49          Y     1029       0.333333   

     new_cell              area creditcd_y  
0           U       NEW ENGLAND          Y  
1           U    ATLANTIC SOUTH          N  
2           U     NEW YORK CITY          Y  
3           U           CHICAGO          Y  
4           U     NORTH FLORIDA          N  
...       ...               ...        ...  
9995        U           CHICAGO          Y  
9996        U         SOUTHWEST          Y  
9997        Y       NEW ENGLAND          Y  
9998        U  CALIFORNIA NORTH          Y  
9999        U       LOS ANGELES          Y  

[10000 rows x 9 columns]

In [81]:
DF_Bien=pd.merge(df,agr_df,how='inner',left_on='Customer_ID',right_on='Customer_ID')
DF_Bien

Customer_ID   totmou   totrev creditcd_x  eqpdays  custcare_Mean  \
0         1082423   7269.0   906.58          Y      371      14.333333   
1         1069232  13552.0  1172.25          N       83       3.666667   
2         1031705   4152.0   577.39          Y      675       3.000000   
3         1089806   1863.0   294.26          Y      218      11.000000   
4         1076695   4155.0   433.16          N      222       0.000000   
...           ...      ...      ...        ...      ...            ...   
9992      1027614   2570.0   484.31          Y      611       0.666667   
9993      1098422   2105.0   583.36          Y     1243       0.000000   
9994      1096497   3690.0   294.29          Y      212       0.666667   
9995      1087248   3333.0   233.08          Y      290       0.000000   
9996      1099003   5025.0  1436.49          Y     1029       0.333333   

     new_cell              area creditcd_y  
0           U       NEW ENGLAND          Y  
1           U    ATLANTIC SOUTH          N  
2           U     NEW YORK CITY          Y  
3           U           CHICAGO          Y  
4           U     NORTH FLORIDA          N  
...       ...               ...        ...  
9992        U           CHICAGO          Y  
9993        U         SOUTHWEST          Y  
9994        Y       NEW ENGLAND          Y  
9995        U  CALIFORNIA NORTH          Y  
9996        U       LOS ANGELES          Y  

[9997 rows x 9 columns]

In [82]:
df3=df.merge(agr_df,how='left',left_on=['creditcd','Customer_ID'],right_on=['creditcd','Customer_ID'])
df3

Customer_ID   totmou   totrev creditcd  eqpdays  custcare_Mean new_cell  \
0         1082423   7269.0   906.58        Y      371      14.333333        U   
1         1069232  13552.0  1172.25        N       83       3.666667        U   
2         1031705   4152.0   577.39        Y      675       3.000000        U   
3         1089806   1863.0   294.26        Y      218      11.000000        U   
4         1076695   4155.0   433.16        N      222       0.000000        U   
...           ...      ...      ...      ...      ...            ...      ...   
9995      1027614   2570.0   484.31        Y      611       0.666667        U   
9996      1098422   2105.0   583.36        Y     1243       0.000000        U   
9997      1096497   3690.0   294.29        Y      212       0.666667        Y   
9998      1087248   3333.0   233.08        Y      290       0.000000        U   
9999      1099003   5025.0  1436.49        Y     1029       0.333333        U   

                  area  
0          NEW ENGLAND  
1       ATLANTIC SOUTH  
2        NEW YORK CITY  
3              CHICAGO  
4        NORTH FLORIDA  
...                ...  
9995           CHICAGO  
9996         SOUTHWEST  
9997       NEW ENGLAND  
9998  CALIFORNIA NORTH  
9999       LOS ANGELES  

[10000 rows x 8 columns]

In [83]:
df3=df.merge(agr_df,left_on='Customer_ID',right_on='Customer_ID',how="inner")
df3
df3[df3.isnull().any(axis=1)]
df3['area']=[i.replace('AREA.csv','').replace('Bases/','') for i in df3['area']]
df3

Customer_ID   totmou   totrev creditcd_x  eqpdays  custcare_Mean  \
0         1082423   7269.0   906.58          Y      371      14.333333   
1         1069232  13552.0  1172.25          N       83       3.666667   
2         1031705   4152.0   577.39          Y      675       3.000000   
3         1089806   1863.0   294.26          Y      218      11.000000   
4         1076695   4155.0   433.16          N      222       0.000000   
...           ...      ...      ...        ...      ...            ...   
9992      1027614   2570.0   484.31          Y      611       0.666667   
9993      1098422   2105.0   583.36          Y     1243       0.000000   
9994      1096497   3690.0   294.29          Y      212       0.666667   
9995      1087248   3333.0   233.08          Y      290       0.000000   
9996      1099003   5025.0  1436.49          Y     1029       0.333333   

     new_cell              area creditcd_y  
0           U       NEW ENGLAND          Y  
1           U    ATLANTIC SOUTH          N  
2           U     NEW YORK CITY          Y  
3           U           CHICAGO          Y  
4           U     NORTH FLORIDA          N  
...       ...               ...        ...  
9992        U           CHICAGO          Y  
9993        U         SOUTHWEST          Y  
9994        Y       NEW ENGLAND          Y  
9995        U  CALIFORNIA NORTH          Y  
9996        U       LOS ANGELES          Y  

[9997 rows x 9 columns]

### Ejercicio 2  
¿Cuáles son los 3 registros que no se incluyen en df3 ?


In [85]:
DF_Full[DF_Full['totmou'].isna()]

Empty DataFrame
Columns: [Customer_ID, totmou, totrev, creditcd_x, eqpdays, custcare_Mean, new_cell, area, creditcd_y]
Index: []

In [86]:
DF_Full[DF_Full['area'].isna()]

Customer_ID   totmou   totrev creditcd_x  eqpdays  custcare_Mean  \
3381      1050270   1409.0   407.02          Y      102            NaN   
4281      1020629  15284.0  2370.62          N      445            NaN   
7384      1020451   3604.0   662.42          Y      675            NaN   

     new_cell area creditcd_y  
3381      NaN  NaN        NaN  
4281      NaN  NaN        NaN  
7384      NaN  NaN        NaN

Si deseamos realizar una filtración sobre algún valor relevante por ejemplo el área podremos revertir el proceso usando el operador `==`  que retorna `True` en el caso que dos objetos tengan el mismo valor

In [87]:
DF_Bien.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9997 entries, 0 to 9996
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Customer_ID    9997 non-null   int64  
 1   totmou         9997 non-null   float64
 2   totrev         9997 non-null   float64
 3   creditcd_x     9997 non-null   object 
 4   eqpdays        9997 non-null   int64  
 5   custcare_Mean  9997 non-null   float64
 6   new_cell       9997 non-null   object 
 7   area           9997 non-null   object 
 8   creditcd_y     9845 non-null   object 
dtypes: float64(3), int64(2), object(4)
memory usage: 703.0+ KB


In [88]:
df3=DF_Bien

In [89]:
new_england=df3[df3["area"]=="NEW ENGLAND AREA"]
new_england


Empty DataFrame
Columns: [Customer_ID, totmou, totrev, creditcd_x, eqpdays, custcare_Mean, new_cell, area, creditcd_y]
Index: []

El comando   `df3["area"]=="NEW ENGLAND AREA"`

In [93]:
df3["area"]=="NEW ENGLAND"

0        True
1       False
2       False
3       False
4       False
        ...  
9992    False
9993    False
9994     True
9995    False
9996    False
Name: area, Length: 9997, dtype: bool

devuelve una serie boleana  con el mismo número de filas de la base`df3`, donde cada valor es `True` o `False` dependiendo si el valor de `area` de un registro  toma el valor de "NEW ENGLAND AREA" 

In [92]:
df3[df3["area"]=="NEW ENGLAND"]

Customer_ID   totmou   totrev creditcd_x  eqpdays  custcare_Mean  \
0         1082423   7269.0   906.58          Y      371      14.333333   
10        1076771    288.0   374.98          Y      284       0.000000   
12        1083737   1928.0   396.78          Y      261       0.000000   
17        1078557   2781.0   590.26          Y      276       0.000000   
18        1018015  24069.0  2129.74          Y      199       2.000000   
...           ...      ...      ...        ...      ...            ...   
9898      1059454    942.0   625.98          Y      323       0.000000   
9900      1036973   2538.0   981.41          Y      688       0.000000   
9959      1073829   2812.0   454.64          Y      342       0.000000   
9990      1083458  10628.0   965.57          Y      262       6.666667   
9994      1096497   3690.0   294.29          Y      212       0.666667   

     new_cell         area creditcd_y  
0           U  NEW ENGLAND          Y  
10          U  NEW ENGLAND          Y  
12          U  NEW ENGLAND          Y  
17          N  NEW ENGLAND          Y  
18          N  NEW ENGLAND          Y  
...       ...          ...        ...  
9898        U  NEW ENGLAND          Y  
9900        U  NEW ENGLAND          Y  
9959        U  NEW ENGLAND          Y  
9990        U  NEW ENGLAND          Y  
9994        Y  NEW ENGLAND          Y  

[525 rows x 9 columns]

### Agrupación por área geográfica

En el Analisis de la información se suelen usar funciones de resumen, las cuales plantean el agrupamiento de ciertas características por categorías. Para más información consultar [gropby()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html).

Para nuestro desarrollo es de interés el revisar si el comportamiento de las variables se mantiene en cada una de las áreas geográficas.


In [ ]:
f_lam=lambda x,y: x+y 

In [ ]:
f_lam(2,3)

In [ ]:
df3[df3['area']=='CALIFORNIA NORTH']['creditcd_x'].value_counts(dropna=False).index

In [ ]:
Agrupacion=df3.groupby("area").agg({'Customer_ID':'count','totmou':['sum','mean'],'creditcd_x':[(lambda x : x.value_counts().Y),(lambda x : x.value_counts().N)]})
Agrupacion.columns=['Número de clientes','Consumo total de minutos','Consumo Promedio Minutos','Con Credito','Sin Credito']

In [ ]:
Agrupacion

Aquí, el objeto DataFrameGroupBy  contiene un objeto DataFrame para cada grupo (en este caso, un objeto DataFrame para cada símbolo). Específicamente, cada elemento del objeto es una tupla que contiene el identificador de grupo (en este caso el símbolo), y las filas correspondientes del DataFrame que tienen ese símbolo).

La librería `pandas`  permite iterar sobre el objeto groupby() para ver lo que hay dentro:

In [ ]:
group_df = df3.groupby(["area"]) # Datos del grupo en df3 filtrados por el área
# Haciendo un bucle a través de los grupos
for item in group_df:
    print(type(item)) # Mostrando el tipo de artículo en df3
    print(item[0]) # área
    print(item[1].head()) # DataFrame

Usando el método  `groupby()` con el método `describe()` y apliquémoslo a cada símbolo para
analizar la distribución de las características la variable ----.

In [ ]:
group_df = df3.groupby(["area"]).describe()
group_df['totmou','mean']

En `pandas`tenemos una manera más sencilla de mostrar estos resultados

In [ ]:
df3[["area","totrev"]].groupby("area").describe()

Estos resultados son iguales a los realizados con el ciclo `for`. Los objetos de tipo `DataFrameGroupBy` permiten una rápida salida de generación de estadísticas para grupos de interés.

### Ejercicio 3

Realice una agrupación por la variable creditcd_x y encuentre el valor de la desviación estándar

### Etiquetando valores de total de ingresos en categorías

Revisamos el comportamiento de `totrev` y encontramos una gran variabilidad, en ocasiones la agrupación de variables cuantitativas en diferentes categorías nos ayuda a tener una mejor visión del conjunto de datos. 

Crearemos una nueva variable `Nivel_totrev`  que catalogue a la variable total de ingresos `totrev` en 2 posibles niveles, alto en el caso que `totrev`  sea mayor al cuantil 50 y baja en otro caso. cada valor de percentil debe ser calculado por cada `area` para asegurar que el +area geografíca sea evaluada adicionalmente.



In [ ]:
pencentiles50 = df3.groupby("area")["totrev"].quantile(0.5) # percentil 50
print(pencentiles50)

Como nos gustaría etiquetar los ingreos en alto y bajopara  cada área geográfica, haremos uso del método
`np.where()` en la biblioteca numpy. Este método toma una entrada y comprueba una condición lógica: si la
condición es verdadera, devolverá su segundo argumento, mientras que si la condición es falsa, devolverá su
tercer argumento. 


In [ ]:
# Loop a través de los símbolos
areas = [i.replace('AREA.csv','').replace('Bases/','') for i in df3['area'].unique()]# registro de las áreas
lista_df = []
# ciclo sobre todos los símbolos
for i in areas:
    print("Etiqueta por área: " + i)
    temp_df = df3[df3["area"] == i] .copy()
    umbral_punto = pencentiles50.loc[i]
    temp_df["Nivel_totrev"] = np.where(temp_df["totrev"] < umbral_punto, "Bajo", "Alto") 
    lista_df.append(temp_df)
df_con_etiquetas = pd.concat(lista_df)

In [ ]:
df_con_etiquetas.head()

In [ ]:
df_con_etiquetas.head()

Ahora podemos hacer una evaluaciónd de como es el comportamiento de los minutos gastados y el nivel de ingresos de los clientes

### ¿Existe relación alguna entre el total de minutos usados con la compañía y el ingreso total de los clientes?

Para explorar la relación entre el nivel de total de ingresos y el número promedio de llamadas, agrupemos por
Nivel_totrev y miremos ingreso total promedio por cada +area geográfica.



In [ ]:
df_con_etiquetas.groupby(['area','Nivel_totrev'])[['totmou']].mean()

### Ejercicio 4


Escriba el código para categorizar  el total de ingresos  baja, media y alta volatilidad, donde:

`
si totrev > (percentil 75 de totrev para el área dada):
Nivel_totrev = 'Alto'
o si VolStat > (percentil 25 de totrev para el área dada):
Nivel_totrev  = 'Medio'
de lo contrario:
Nivel_totrev  = 'Bajo'`

In [ ]:
pencentiles25 = df3.groupby("area")["totrev"].quantile(0.25) # percentil 50
pencentiles75 = df3.groupby("area")["totrev"].quantile(0.75)

In [ ]:
# registro de las áreas
lista_df = []
# ciclo sobre todos los símbolos
for i in areas:
    print("Etiqueta por área: " + i)
    temp_df = df3[df3["area"] == i] .copy()
    umbral_punto1 = pencentiles25.loc[i]
    umbral_punto2 = pencentiles75.loc[i]
    lv=[]
    for i in temp_df['totrev']:
        if i<umbral_punto1:
            lv.append('Bajo')
        elif i<umbral_punto2:
            lv.append('Medio')
        else:
            lv.append('Alto')
    temp_df['Nivel']=lv
    lista_df.append(temp_df)
df_con_etiquetas = pd.concat(lista_df)
# realice el ejercicio en este espacio

In [ ]:
df_con_etiquetas

In [ ]:
df_con_etiquetas.groupby(['area','Nivel'])[['totmou']].mean()

### Visualización de total de llamadas y total de ingresos

Ya hemos respondido satisfactoriamente a nuestra pregunta original. Sin embargo, no es necesario solamente
analizar los datos en formato tabular. Python contiene una funcionalidad que le permite analizar sus datos
visualmente también.

Usaremos la funcionalidad de pandas sobre la librería estándar de graficación de Python, `matplotlib`. Vamos
a importar la librería e instruir a Jupyter que muestre los gráficos en línea (es decir, mostrar los gráficos en
la pantalla del cuaderno para que podamos verlos mientras ejecutamos el código):



In [ ]:
modules = dir()
print(modules)

In [ ]:
import matplotlib.pyplot as plt
# Graficar en el cuaderno
%matplotlib inline

#### Histograma

In [ ]:
plt.style.use('ggplot')
plt.hist(df3["totrev"],bins=13,color="#237AAF")
plt.title('Histograma de variable totrev')
plt.xlabel("totrev");plt.ylabel("Frecuencia")
plt.show()

In [ ]:
print(plt.style.available)

#### Boxplot

In [ ]:
plt.style.use('ggplot')
plt.boxplot(df3["totrev"])
plt.title('Boxplot de variable totrev')
plt.show()

#### Gráficas a partir de `groupby`

In [ ]:
df4=df_con_etiquetas.groupby(['area','Nivel'])[['totmou']].mean()
df4.plot(kind='barh',figsize=(10,8))
plt.show()

#### Gráficas de `pandas`  

El modulo`.plot()` genera por defecto una gráfica de línea, existen varios gráficos para los `DataFrameGroup`

    - 'line' : Gráfica de línea
    - 'bar' :  Diagrama de barras vertical
    - 'barh' : Diagrama de barras horizontal
    - 'hist' : histograma
    - 'box' : boxplot
    - 'kde' : Kernel de densidad
    - 'area' : Gráfica de área
    - 'pie' : Diagrama de torta
    - 'scatter' : Diagrama de dispersión
    - 'hexbin' : Diagrama de hexágonos

In [ ]:
df3.columns

In [ ]:
df3["creditcd_x"].value_counts().plot(kind="pie")
plt.show()

Revisando la relación entre las dos variables `totrev` y `totmou`

In [ ]:
df3.plot(kind="scatter",x='totrev',y='totmou',c='blue')

### `Seaborn`

In [ ]:
#!pip install seaborn

In [ ]:
import seaborn as sns
sns.boxplot(df3["totrev"],orient='v')
plt.title('Boxplot de variable totrev')
plt.show()

### Gráficas de parcela 

In [ ]:
df3

In [ ]:
plt.figure(figsize=(116,4))
sns.pairplot(data=df3, hue='creditcd_x', vars=['totrev','totmou','eqpdays'])
plt.title("Parcelas")
plt.show()

In [ ]:
plt.figure(figsize=(6,4))
sns.boxplot(x='Nivel',y='totmou',data=df_con_etiquetas)
plt.title('Boxplot múltiple para Nivel de ingresos totales y \n total de minutos gastados')
plt.show()

In [ ]:
sns.catplot(y="totmou", x="Nivel", kind="box",    data=df_con_etiquetas, height=4, aspect=4, palette='Set2')
sns.catplot(y="totmou", x="Nivel", kind="violin", data=df_con_etiquetas, height=4, aspect=4, palette='Set2')

In [ ]:
ax = sns.catplot(y="totmou", x="Nivel", hue='creditcd_x', kind='point',data=df_con_etiquetas, aspect=2)
ax = sns.catplot(x="Nivel", kind='count',data=df_con_etiquetas, col='creditcd_x')

### Ejercicio 5

Realice un  boxplot para la variable totmou segmentado por las variable  credict_x y Nivel_totrev

In [ ]:
# realice el ejercicio en este espacio

## Resumen


1. Leer datos desde archivos.
2. Agregar y manipular datos .
3. Analizar estadísticas de resumen y reunir información.
4. Usar matplotlib para crear gráficos para hacer análisis visual, además de otras librerías.

Para más información

> [https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html](https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html)